### PR TITLE
Fix boost versions in docs

### LINKED ISSUE #13870: Required boost version in the docs does not match the code after recent changes
## Page

https://docs.soliditylang.org/en/latest/installing-solidity.html#building-from-source

## Abstract

According to the above page, for C++ we have as dependencies: 
       Boost (version 1.77+ on Windows, 1.65+ otherwise) | C++ libraries.

with a note below: "Starting from 0.5.10 linking against Boost 1.70+ should work without manual intervention."

But in this change: https://github.com/ethereum/solidity/commit/1da82045c4881e5ccd6f15f7abf472272e1028cd "fs::directory_options" was used. a feature which was added in boost 1.72.

What's the minimum boost version you are targeting?

### PR DESCRIPTION
**Pull request by nikola-matic**
Fixes https://github.com/ethereum/solidity/issues/13870

### CODE REVIEW

File: docs/installing-solidity.rst [hunk 1]
```diff
@@ -338,11 +338,11 @@ The following are dependencies for all builds of Solidity:
 .. _z3: https://github.com/Z3Prover/z3
 
 .. note::
```
Comment (nikola-matic) [inline docs/installing-solidity.rst:340]: The question for this whole `note` section is - do we leave it as is (1.74), remove it altogether, or let it stay as it was (1.70)?

Comment (ekpyron) [reply docs/installing-solidity.rst:340]: Maybe we just remove it entirely - we can't keep acurately documenting builds of arbitrarily old versions here anyways.

Comment (nikola-matic) [reply docs/installing-solidity.rst:340]: Agreed.

**Comment by cameel** [comment]
The check in `CMakeLists.txt` needs to be updated to match: https://github.com/ethereum/solidity/blob/609f15229bd72310925b038ee83328cab13c3043/cmake/EthDependencies.cmake#L34-L41

Please also remember to include a short comment pointing out why earlier versions won't work.

**Comment by cameel** [comment]
There are also two more places that check version, but we probably can leave them at 1.65:
https://github.com/ethereum/solidity/blob/609f15229bd72310925b038ee83328cab13c3043/libsolutil/Numeric.h#L28-L30
https://github.com/ethereum/solidity/blob/609f15229bd72310925b038ee83328cab13c3043/libsolutil/Common.h#L44-L46

They were added in #5824, apparently because earlier versions required some workarounds for multiprecision. So this particular code should still work on 1.65.

File: cmake/EthDependencies.cmake [hunk 2]
```diff
@@ -34,10 +34,9 @@ set(BOOST_COMPONENTS "filesystem;unit_test_framework;program_options;system")
 if (WIN32)
 	# Boost 1.77 fixes a bug that causes crashes on Windows for some relative paths in --allow-paths.
 	# See https://github.com/boostorg/filesystem/issues/201
-	find_package(Boost 1.77.0 QUIET REQUIRED COMPONENTS ${BOOST_COMPONENTS})
+	find_package(Boost 1.77.0 EXACT QUIET REQUIRED COMPONENTS ${BOOST_COMPONENTS})
```
Comment (ekpyron) [inline cmake/EthDependencies.cmake:38]: Is there really call for this? How bad are the issues with > 1.78? But well, ok, maybe we can keep it for now, since we have an issue staged for fixing it and can then relax it again...

File: cmake/EthDependencies.cmake [hunk 3]
```diff
@@ -34,10 +34,9 @@ set(BOOST_COMPONENTS "filesystem;unit_test_framework;program_options;system")
...
 else()
-	# Boost 1.65 is the first to also provide boost::get for rvalue-references (#5787).
-	find_package(Boost 1.65.0 QUIET REQUIRED COMPONENTS ${BOOST_COMPONENTS})
+	find_package(Boost 1.74.0 QUIET REQUIRED COMPONENTS ${BOOST_COMPONENTS})
```
Comment (cameel) [inline cmake/EthDependencies.cmake]: Please mention why 1.78+ does not work on Windows and why versions earlier than 1.65 are too old.

**Review by cameel — DISMISSED** [review]
Looks fine, but comments explaining the requirement would be nice.

File: cmake/EthDependencies.cmake (hunk 2 above)
Comment (nikola-matic) [reply cmake/EthDependencies.cmake:38]: 1. Unless you upgrade to visual studio 2022 (2020 is documented as working), you can't even build 1.78
2. If you do upgrade and get everything running - soltest fails due to different behaviour of `boost::filesystem`

File: cmake/EthDependencies.cmake (hunk 3 above)
Comment (nikola-matic) [reply cmake/EthDependencies.cmake]: For Windows - sure. For 1.65 I don't really see how it makes sense to document versions <1.65 if the minimum we require is 1.74? Also, the reason we're supporting 1.74 is because that's the minimum version used by our CI, and we have no intention of supporting boost versions that we don't use.

**Comment by nikola-matic** [comment]
> There are also two more places that check version, but we probably can leave them at 1.65:
> 
> https://github.com/ethereum/solidity/blob/609f15229bd72310925b038ee83328cab13c3043/libsolutil/Numeric.h#L28-L30
> 
> https://github.com/ethereum/solidity/blob/609f15229bd72310925b038ee83328cab13c3043/libsolutil/Common.h#L44-L46
> 
> They were added in #5824, apparently because earlier versions required some workarounds for multiprecision. So this particular code should still work on 1.65.

If the current build passes, I'll make the changes here as well.

File: cmake/EthDependencies.cmake (hunk 2 above)
Comment (ekpyron) [reply cmake/EthDependencies.cmake:38]: Yeah, but how bad are the ``boost::filesystem`` failures? Is that mainly technicalities or proper issues? Depending on that, I'd just allow >1.77 here anyways... overrestricting the build system requirements is almost as bad as underrestricting them...

Comment (nikola-matic) [reply cmake/EthDependencies.cmake:38]: We already have an issue open to bump everything on Windows, I believe Chris P tried it some time ago, then I tried it a couple of months ago, and in general I feel as restricting it to an exact version is the correct move until we fully fix this - but I can certainly remove the `EXACT` from here since no one really builds Solidity from source on Windows anyway :D

Comment (ekpyron) [reply cmake/EthDependencies.cmake:38]: IIRC, MSVC2022 kind of works if boost is updated (I think I did it myself, but I don't recall precisely). And the only issues with newer boost are rather exotic cases (but not sure about that)... so not sure if it's wise to "fix" this by making an MSVC2022 build downright impossible, whereas without this it may just be a hassle with merely technically incorrect corner case behaviour...

Comment (ekpyron) [reply cmake/EthDependencies.cmake:38]: But yeah, ideally we'd just properly fix this anyways. Ideally we'd also just move to ``std::filesystem``, we just need to check if os x is still messed up in that regard... but neither of that we can do for this release in any case...
But I'd still argue against fixing ``EXACT`` here for the release... IMHO, that does more harm than good...

Comment (nikola-matic) [reply cmake/EthDependencies.cmake:38]: > IIRC, MSVC2022 kind of works if boost is updated (I think I did it myself, but I don't recall precisely). 

It's the other way around - you need MSVC2022 in order to build boost 1.78 (`b2` )
I'll remove the `EXACT` as you make a good enough point :)

Comment (cameel) [reply cmake/EthDependencies.cmake:38]: I don't think many users would be affected in practice (or at least they may not notice) but it changes compiler behavior depending on which Boost version you compile with so I'd still consider it highly undesirable. In corner cases you could get different metadata between platforms for the same code, depending on how you build the compiler.

From what we established in #13434, the normalization on newer versions is different for paths starting with `//`. And in some cases you might get stuff normalized to `\path` instead of `\path\.` as before. The latter probably does not matter much for us (I'd have to make sure though) but you could run into the former if you're sloppy about your paths.

Overall the fact that a part of the language depends on aspects of `filesystem` that can and do change between versions (and implementations - `std::filesystem` has slightly different rules as well) is messy. You think you're saving yourself a lot of hassle by not having to write all that FS code but then you have to have a very detailed test suite to make sure behavior stays consistent. And in our case not all of our filesystem use is covered with tests so I would not be surprised if there are more things like this that we just do not see.

**Comment by cameel** [comment]
> If the current build passes, I'll make the changes here as well.

I would not really bump that version. I'd consider just removing it because the one in CMakeFiles.txt covers this anyway. But I'm also fine with leaving it as is. It does seem accurate for these particular files.

**Comment by nikola-matic** [comment]
> > If the current build passes, I'll make the changes here as well.
> 
> I would not really bump that version. I'd consider just removing it because the one in CMakeFiles.txt covers this anyway. But I'm also fine with leaving it as is. It does seem accurate for these particular files.

I'll remove it.

File: cmake/EthDependencies.cmake (hunk 3 above)
Comment (cameel) [reply cmake/EthDependencies.cmake]: Ah, sorry, I meant 1.74.

> Also, the reason we're supporting 1.74 is because that's the minimum version used by our CI, and we have no intention of supporting boost versions that we don't use.

Sure. Just put this in a comment. The fact that it's completely arbitrary is even more reason to do it. We have this info at hand now but later can be very frustrating dig out the reason and figure out if it's fine to change it or not.

File: cmake/EthDependencies.cmake (hunk 2 above)
Comment (cameel) [reply cmake/EthDependencies.cmake:38]: I think that `EXACT` should stay. Build with 1.78 simply gives you the compiler that does not behave correctly, even if only in corner cases. I mean, as a user, why would you want to use version known to be even slightly broken if you can get a perfectly good one with 1.77?

The fact that tests won't pass will give you hint that something's wrong but that's too late. User is likely to ignore it after going through all the hassle of installing all dependencies and getting it to build. It's much more helpful to fail immediately.

File: docs/installing-solidity.rst [hunk 4]
```diff
@@ -321,8 +321,8 @@ The following are dependencies for all builds of Solidity:
 +===================================+=======================================================+
 | `CMake`_ (version 3.13+)          | Cross-platform build file generator.                  |
```
Comment (NunoFilipeSantos) [inline docs/installing-solidity.rst:322]: Shouldn't this change to CMake 3.21.3 for 1.77?

Comment (NunoFilipeSantos) [reply docs/installing-solidity.rst:322]: When you have a new Boost release, this information is added on the next release on CMake. So it's always the CMake version right after the Boost release date.

Comment (NunoFilipeSantos) [reply docs/installing-solidity.rst:322]: Or at least it used to be this way.

File: cmake/EthDependencies.cmake (hunk 2 above)
Comment (nikola-matic) [reply cmake/EthDependencies.cmake:38]: I'll put the `EXACT` back since I have to rebase anyway.

File: docs/installing-solidity.rst (hunk 4 above)
Comment (nikola-matic) [reply docs/installing-solidity.rst:322]: Haven't seen anything in CMake changelogs that references boost, and am not really sure why a boost upgrade would require a CMake bump as well (unless the boost directory structure or something similar changes). `3.13` is quite old, so perhaps we could bump that too though. @cameel @ekpyron Also @r0qs what version of CMake do we use on our ubu2204 images?

Comment (ekpyron) [reply docs/installing-solidity.rst:322]: It at the very least used to be true - i.e. the cmake files for finding boost were gloriously shipped with cmake itself not with boost, so if your boost was newer than your cmake, things broke...
IIRC they wanted to fix that, but not sure that's done just yet.

File: cmake/EthDependencies.cmake (hunk 2 above)
Comment (ekpyron) [reply cmake/EthDependencies.cmake:38]: > I think that `EXACT` should stay. Build with 1.78 simply gives you the compiler that does not behave correctly, even if only in corner cases. I mean, as a user, why would you want to use version known to be even slightly broken if you can get a perfectly good one with 1.77?
> 

Because you can't build against 1.77 at all with MSVC2022, but you can with 1.78, I think. I'm rather sure something broke otherwise when updating MSVC, but maybe I'm misremembering.

Comment (nikola-matic) [reply cmake/EthDependencies.cmake:38]: Ultimately, bumping to 1.78 requires MSVC2022, and if both are bumped, our builds do not succeed (bytecode compare if remember correctly), which I believe is not good enough for us to recommend it in the docs.

File: docs/installing-solidity.rst (hunk 4 above)
Comment (ekpyron) [reply docs/installing-solidity.rst:322]: I mentioned it before, but as a rule of thumb, I wouldn't bump minimum requirements beyond what current debian stable has without good reason - that'd be cmake 3.25.1...

Just to be clear: we're currently dropping support for building on vanilla ubuntu 20.04? What exactly would it take to be able to gracefully build there *and* on 22.04? If that was possible, I'd prefer it, but not if we have to get out of our way for it (I didn't follow the discussion last week precisely, since I was actually off when you bumped this).

Comment (ekpyron) [reply docs/installing-solidity.rst:322]: If we drop support for building on 20.04, we have to additionally drop ``focal`` from ``scripts/release_ppa.sh`` and add it to the list of versions to which we copy the static binaries on launchpad in our release checklist.

Comment (cameel) [reply docs/installing-solidity.rst:322]: Yeah, I also think we should keep the requirement low enough to keep it working on distros people still use. I think Ubuntu 20.04 is still reasonable.

By the way, I actually wanted to say that we still have buildpack images for even older Ubuntu so it should be working but apparently not. The ossfuzz image is named after 16.04 but looking at the source, [it seems to be actually using 20.04](https://github.com/google/oss-fuzz/blob/master/infra/base-images/base-image/Dockerfile#L19). Should we rename it (not in this PR of course)?

File: cmake/EthDependencies.cmake (hunk 2 above)
Comment (ekpyron) [reply cmake/EthDependencies.cmake:38]: Yes, but also the other way round. I'm categorically against fixing versions in cmake.

Resp. to be clear:
- bumping boost to 1.78 requires bumping MSVC to 2022
- bumping MSVC to 2022 requires bumping boost to 1.78
- The builds succeed with boost 1.78 and MSVC 2022, but have minor hickups in ``boost::filesystem``

File: docs/installing-solidity.rst (hunk 4 above)
Comment (ekpyron) [reply docs/installing-solidity.rst:322]: But yeah, IIRC now, the issue was that osx/homebrew bumped the boost version, causing builds to fail, and supporting both older and newer boost was tricky, so we bumped the minimum version, effectively dropping ubuntu 20.04 support, right? That'd be fine, but yeah, that'd mean we have to do https://github.com/ethereum/solidity/pull/13874#discussion_r1071525543, otherwise our release will fail.

Comment (cameel) [reply docs/installing-solidity.rst:322]: Right. Fine with dropping it then.

File: cmake/EthDependencies.cmake (hunk 2 above)
Comment (cameel) [reply cmake/EthDependencies.cmake:38]: Well, maybe I'll just try to create a proper fix for the boost changes tomorrow. They are not actually that big so maybe we can avoid this problem by just making it work with both 1.77 and 1.78.

Comment (nikola-matic) [reply cmake/EthDependencies.cmake:38]: > Well, maybe I'll just try to create a proper fix for the boost changes tomorrow. They are not actually that big so maybe we can avoid this problem by just making it work with both 1.77 and 1.78.

This was supposed to be a 5 minute doc change PR, and it's already turned into 36 comment discussion. I honestly don't think this should be our priority at the moment, especially seeing as we haven't released in a while.

File: docs/installing-solidity.rst (hunk 4 above)
Comment (ekpyron) [reply docs/installing-solidity.rst:322]: Haha, yeah, I was wondering about the fuzzing image as well - we should probably just drop the ubuntu version from it altogether.

Comment (ekpyron) [reply docs/installing-solidity.rst:322]: But yeah, since the only backwards-incompatible change in https://github.com/ethereum/solidity/pull/13866 seems to be the lsp change, which we can drop eventually anyways, I wonder if we should just ``#ifdef`` that after all and could keep supporting 20.04 that way (would be enough to confirm build and tests locally without a full-blown CI run)...

File: cmake/EthDependencies.cmake (hunk 2 above)
Comment (ekpyron) [reply cmake/EthDependencies.cmake:38]: We shouldn't even have started messing with the unrelated windows case here in the first place. But the other thread https://github.com/ethereum/solidity/pull/13874#discussion_r1071488393 here is rather important.

File: docs/installing-solidity.rst (hunk 4 above)
Comment (ekpyron) [reply docs/installing-solidity.rst:322]: I.e. there's no call for dropping 20.04 in favour of a single ``#ifdef`` in code we're likely to drop at the next opportunity anyways :-).

Comment (NunoFilipeSantos) [reply docs/installing-solidity.rst:322]: Going back to my original question, I _really_ think the CMake version on the table would be incompatible if you make the Boost 1.77 min version. it needs to be any CMake version released after Boost 1.77.

**Comment by nikola-matic** [comment]
Closing in lieu of https://github.com/ethereum/solidity/pull/13880
